### PR TITLE
[#IA] Tests for register hospital on pharmacy

### PR DESCRIPTION
**Pull request by maja-tepavcevic**

### CODE REVIEW

File: IntegrationAdapters/Migrations/20201123171708_FirstMigration.Designer.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Infrastructure;
+using Microsoft.EntityFrameworkCore.Migrations;
+using Microsoft.EntityFrameworkCore.Storage.ValueConversion;
+
+namespace IntegrationAdapters.Migrations
+{
+    [DbContext(typeof(HealthCareSystemDbContext))]
+    [Migration("20201123171708_FirstMigration")]
+    partial class FirstMigration
```
Comment (milicasiriski) [inline IntegrationAdapters/Migrations/20201123171708_FirstMigration.Designer.cs]: You can delete this whole folder after the migration is done 😄

File: IntegrationAdapters/Models/Api.cs [hunk 2]
```diff
@@ -6,15 +6,22 @@
...
     public class Api
     {
         [Key]
         public string Key { get; set; }
         public string PharmacyName { get; set; }
         public string Url { get; set; }
 
-        public Api() {}
+        public Api() { }
+
```
Comment (milicasiriski) [inline IntegrationAdapters/Models/Api.cs]: one white space is enough

File: IntegrationAdaptersTests/RegisterHospitalTests.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    {
+        public readonly HealthCareSystemDbContext dbContext;
+        [Fact]
+        public void Is_pharmacy_already_existing_on_hospital()
+        {
+            var options = new DbContextOptionsBuilder<HealthCareSystemDbContext>()
+                .UseMySql(connectionString: "server=localhost;port=3306;database=newmydb;user=root;password=root")
+                .Options;
+
+            ApiService service = new ApiService(new HealthCareSystemDbContext(options));
```
Comment (milicasiriski) [inline IntegrationAdaptersTests/RegisterHospitalTests.cs]: I think the convention is to call the object we are testing 'subject' which is short for 'subject under test' so that we emphasize what we are testing. Also options object can be extracted as a field because it's used in both tests - this will make the tests shorter and easier to maintain.

Comment (milicasiriski) [reply IntegrationAdaptersTests/RegisterHospitalTests.cs]: Maybe you should think about using mocks in these tests instead of the real data base - that way the Apis you are using aren't dependent on the actual data in the database? Also you should add tests for pharmacy registration

**Review by milicasiriski — CHANGES_REQUESTED** [review]
Considering these are the first real tests in the whole project I think they are good. They are well structured, don't have too many expectations and I like that you've covered more outcomes. What we need to consider though is whether to mock or use the real database, since it's an unmanaged dependency.

File: IntegrationAdaptersTests/RegisterHospitalTests.cs (hunk 3 above)
Comment (dragana-todorovic) [reply IntegrationAdaptersTests/RegisterHospitalTests.cs]: Okay, on daily assistant told us it's okay to use real databese because it is integration test, but we need to add unit tests and correct these mistakes. Thanks

File: IntegrationAdaptersTests/RegisterHospitalTests.cs [hunk 4]
```diff
@@ -29,17 +29,51 @@ public void Is_pharmacy_already_existing_on_hospital()
...
+        }
+        [Fact]
+        public void Is_apiKey_not_null()
+        {
+            mock = new Mock<IApiRepository>();
+            mock.Setup(expression: t => t.GetAllApis()).Returns(new List<Api> { new Api { Key = "kljuc1", PharmacyName = "name1", Url = "url1" } });
+            ApiService service = new ApiService();
+            Api api = mock.Object.GetAllApis()[0];
+            Assert.NotNull(service.getApiKey(api));
+        }
```
Comment (milicasiriski) [inline IntegrationAdaptersTests/RegisterHospitalTests.cs]: Is there a business logic case that requires this check?

Comment (milicasiriski) [reply IntegrationAdaptersTests/RegisterHospitalTests.cs]: Formatting tip: Add an empty row between test functions so that they are easier to find 😄 I also like to add an empty line between arrange, act and assert sections to make the test more readable

File: IntegrationAdaptersTests/RegisterHospitalTests.cs [hunk 5]
```diff
@@ -29,17 +29,51 @@ public void Is_pharmacy_already_existing_on_hospital()
...
-                .UseMySql(connectionString: "server=localhost;port=3306;database=newmydb;user=root;password=root")
-                .Options;
+            ApiService service = new ApiService();
 
-            ApiService service = new ApiService(new HealthCareSystemDbContext(options));
-
-            Api a = new Api("m", "m", "m");
+            Api a = new Api("l", "l", "l");
 
             bool notExists = service.RegisterHospitalOnPharmacy(a);
```
Comment (milicasiriski) [inline IntegrationAdaptersTests/RegisterHospitalTests.cs]: A better name for this variable would be registerSuccessful, since the service doesn't know why the hospital couldn't be registered or that api can't be added (the repository contains that logic)

**Review by milicasiriski — COMMENTED** [review]
Good unit tests! Minor refactoring required

File: IntegrationAdaptersTests/RegisterHospitalTests.cs [hunk 6]
```diff
@@ -20,49 +20,32 @@ public void Is_pharmacy_already_existing_on_hospital()
         {          
             ApiService service = new ApiService();
 
-            Api a = new Api("k", "k", "k");
-
-            bool alreadyExists = service.RegisterHospitalOnPharmacy(a);
+            Api registerFailed = new Api("k", "k", "k");
+            bool alreadyExists = service.RegisterHospitalOnPharmacy(registerFailed);
 
             Assert.False(alreadyExists);
```
Comment (milicasiriski) [inline IntegrationAdaptersTests/RegisterHospitalTests.cs]: I would expect that the result of alreadyExists variable is True, so the name is not very fitting, you could just name it result or success

File: IntegrationAdaptersTests/RegisterHospitalTests.cs [hunk 7]
```diff
@@ -20,49 +20,32 @@ public void Is_pharmacy_already_existing_on_hospital()
...
         {
             ApiService service = new ApiService();
 
-            Api a = new Api("l", "l", "l");
-
-            bool notExists = service.RegisterHospitalOnPharmacy(a);
+            Api registerSuccessful = new Api("l", "l", "l");
+            bool notExists = service.RegisterHospitalOnPharmacy(registerSuccessful);
 
             Assert.True(notExists);
```
Comment (milicasiriski) [inline IntegrationAdaptersTests/RegisterHospitalTests.cs]: same here

**Review by milicasiriski — APPROVED** [review]
Everything is fine, just change the variable names to result or success so the tests are more readable 😄 it's a bit confusing now

**Review by dragana-todorovic — APPROVED** [review]
